### PR TITLE
Arrow Key functionality and extracted some movement into functions

### PR DESCRIPTION
**Pull request by RiordanIX**
h, j, k, and l movement logic has been moved into functions so they can
be shared with the arrow keys. The arrow keys work pretty well, there
are just a few cases that need fine tuned (like pressing an arrow key
immediately after pressing escape, that doesn't quite work right).

There are some more escape sequences that can be implemented, like
page up, page down, home, end, etc. I will work on those later.

### CODE REVIEW

**Comment by tomsmeding** [comment]
I really thought I'd implemented input better than `cin.get()`. In [termio](https://github.com/tomsmeding/termio/blob/master/termio.c#L398), and therefore all things after that, I used `select(2)` to get input with a timeout to correctly capture escape sequences etc.
Anyway, this seems fine; but since you have a TODO: open, I'll let you finish that before I merge :)

**Comment by RiordanIX** [comment]
Sounds good, I may be able to get to it today actually.

**Comment by tomsmeding** [comment]
This looks like great work! Can't test right now, but I'll probably look at it tomorrow.
I'm privately busy with a fully rewritten version that has a way better class structure and is based on my termio library, making keyboard input a lot easier. But of course you can continue working on this thing if you want :p

**Comment by RiordanIX** [comment]
Whenever I did it that way, exiting the program kept it in raw mode. It does need to be passed by reference when changing it.

Whenever we declare the termios, like on 131, it creates a copy of the current state of the terminal.  After changing the attributes, you do need to set the attribute by reference to update the current state. So in the disable and enable timeout functions, it gets the current state, modifies the two attributes, then makes that the state of the terminal.

When exiting the application, it resets the terminal to how it was originally, because we never actually changed tios_bak, just copies of it.

**Edit:** I read up on it some more and you're right.  It just so happened that editing still worked after switching the timeouts on the terminal.  Fortunately, this is a two line fix.  We don't really need to backup the slight editing of the terminals, so this should work fine.

**Comment by tomsmeding** [comment]
I do not understand exactly what you're saying above, but that commit is exactly what I meant. :)
And are you aware my comment about the passing by reference is about something totally unrelated?

File: io.cpp [hunk 1]
```diff
@@ -40,6 +40,32 @@ Colour ansiTable[8]={black,red,green,yellow,blue,magenta,cyan,white};
 unsigned int screencolours;
 
 unordered_map<string,string> tputcache;
+enum KEY_ACTION{
+			KEY_NULL = 0,       /* NULL */
+			CTRL_C = 3,         /* Ctrl-c */
```
Comment (tomsmeding) [inline io.cpp:45]: Are you using these constants anywhere?

Comment (RiordanIX) [reply io.cpp:45]: Many of them. They are some specialty keys that can get returned by the `editorReadKey` function. Not all are being used right now.

As for my previous comment, yes, I do realize now that the 'passing by reference' comment was about something different.  I don't remember why I added that comment in the first place, but I can remove it if you'd like.

**Comment by tomsmeding** [comment]
My terminal emits different key codes for things like delete and home/end, so those don't work for me, but because they weren't implemented before in the first place (I believe), this is fine.
I looked through your changes and even though I didn't thoroughly check everything, stuff compiles and seems to work fine, so thumbs up!

If you could apply [this patch](https://gist.github.com/8f5e8950c54c73f8a6071141d77e32ab) and commit it to this PR, that would be awesome. Then I think this is ready to merge.

**Comment by RiordanIX** [comment]
Alright, I added the patch and made a minor change in the Makefile. I realized I was doing the regex all wrong and made it actually work for any version of lua >=5.0, at least on linux.

**Pull request merged**